fem/fe.cpp

```diff
@@ -12627,6 +12637,2047 @@ void ND_SegmentElement::CalcVShape(const IntegrationPoint &ip,
+      }
+   for (int j = 0; j <= p; j++)
+      for (int i = 0; i <= p; i++)
+      {
+         int idx = dof_map[o++];
+         dof2nk[idx] = 4;
+         Nodes.IntPoint(idx).Set2(op[i], op[j]);
+      }
+}
+
+void RT_R2D_QuadrilateralElement::CalcVShape(const IntegrationPoint &ip,
+                                             DenseMatrix &shape) const
+{
+   const int pp1 = order;
+
+#ifdef MFEM_THREAD_SAFE
+   Vector shape_cx(pp1 + 1), shape_ox(pp1), shape_cy(pp1 + 1), shape_oy(pp1);
+#endif
+
+   cbasis1d.Eval(ip.x, shape_cx);
+   obasis1d.Eval(ip.x, shape_ox);
+   cbasis1d.Eval(ip.y, shape_cy);
+   obasis1d.Eval(ip.y, shape_oy);
+
+   int o = 0;
+   for (int j = 0; j < pp1; j++)
+      for (int i = 0; i <= pp1; i++)
+      {
+         int idx, s;
+         if ((idx = dof_map[o++]) < 0)
+         {
+            idx = -1 - idx, s = -1;
+         }
+         else
+         {
+            s = +1;
+         }
+         shape(idx,0) = s*shape_cx(i)*shape_oy(j);
+         shape(idx,1) = 0.;
+         shape(idx,2) = 0.;
+      }
+   for (int j = 0; j <= pp1; j++)
+      for (int i = 0; i < pp1; i++)
+      {
+         int idx, s;
+         if ((idx = dof_map[o++]) < 0)
+         {
+            idx = -1 - idx, s = -1;
+         }
+         else
+         {
+            s = +1;
+         }
+         shape(idx,0) = 0.;
+         shape(idx,1) = s*shape_ox(i)*shape_cy(j);
+         shape(idx,2) = 0.;
+      }
+   for (int j = 0; j < pp1; j++)
+      for (int i = 0; i < pp1; i++)
+      {
+         int idx = dof_map[o++];
+         shape(idx,0) = 0.;
+         shape(idx,1) = 0.;
+         shape(idx,2) = shape_ox(i)*shape_oy(j);
+      }
+}
+
+void RT_R2D_QuadrilateralElement::CalcDivShape(const IntegrationPoint &ip,
+                                               Vector &divshape) const
+{
+   const int pp1 = order;
+
+#ifdef MFEM_THREAD_SAFE
+   Vector shape_cx(pp1 + 1), shape_ox(pp1), shape_cy(pp1 + 1), shape_oy(pp1);
+   Vector dshape_cx(pp1 + 1), dshape_cy(pp1 + 1);
+#endif
+
+   cbasis1d.Eval(ip.x, shape_cx, dshape_cx);
+   obasis1d.Eval(ip.x, shape_ox);
+   cbasis1d.Eval(ip.y, shape_cy, dshape_cy);
+   obasis1d.Eval(ip.y, shape_oy);
+
+   int o = 0;
+   for (int j = 0; j < pp1; j++)
+      for (int i = 0; i <= pp1; i++)
+      {
+         int idx, s;
+         if ((idx = dof_map[o++]) < 0)
+         {
+            idx = -1 - idx, s = -1;
+         }
+         else
+         {
+            s = +1;
+         }
+         divshape(idx) = s*dshape_cx(i)*shape_oy(j);
+      }
+   for (int j = 0; j <= pp1; j++)
+      for (int i = 0; i < pp1; i++)
+      {
+         int idx, s;
+         if ((idx = dof_map[o++]) < 0)
+         {
+            idx = -1 - idx, s = -1;
+         }
+         else
+         {
+            s = +1;
+         }
+         divshape(idx) = s*shape_ox(i)*dshape_cy(j);
+      }
+   for (int j = 0; j < pp1; j++)
+      for (int i = 0; i < pp1; i++)
+      {
+         int idx = dof_map[o++];
+         divshape(idx) = 0.;
+      }
+}
+
+
```

Comment:
This file is getting very long. Should the new code be put in a different `cpp` file?

Reply:
I would support splitting `fe.cpp` and `fe.hpp` into multiple files. Perhaps one for H1, one for ND, one for RT, one for L2, and keep common base classes in `fe.hpp` and `fe.cpp`. Other basis functions could either stay in the current files or be placed in a miscellaneous file.

Reply:
That sounds good to me. We already have similar splittings of the `bilininteg_*` files.